### PR TITLE
feat(Canvas): copy-paste yaml support

### LINKED ISSUE #2652: Copy / Paste YAML support
### Please describe the feature that you want to propose

The Copy / Paste functionality relies on a Kaoto JSON structure to mark the copied element type. While this works, pasting a snippet from another source is not possible.

The goal of this ticket is to remove the KAoto JSON structure and rely on the YAML construct directly.

### PR DESCRIPTION
**Pull request by PVinaches**
Summary

  This PR introduces native YAML clipboard support for copying and pasting visual entities (routes, pipes,
  kamelets, steps) in the Kaoto canvas. Users can now copy entities as YAML and paste them from external sources
  or between different parts of the application.

  ---
  What changed

  Core clipboard functionality

  - Replaced ClipboardManager utility with new ClipboardService
  (packages/ui/src/services/visualization/clipboard.service.ts)
  - Added a new ClipboardService
  (packages/ui/src/services/visualization/clipboard.service.ts)
  - Removed the type field from IClipboardContent interface - clipboard compatibility is now inferred from YAML
  key names instead of metadata
  - Added bidirectional YAML serialization:
    - Copy: Serializes entities to YAML format (top-level entities like route, pipe as objects; steps as arrays
  with leading dash)
    - Paste: Parses YAML from clipboard, accepting both Kaoto format ([{ stepName: {...} }]) and external format
  ({ stepName: {...} })
  - Added dual MIME type support: web text/kaoto (Chrome) and text/plain (all browsers) for better
  interoperability

  Updated components and hooks

  - Updated all visualization hooks (copy-step.hook, paste-step.hook, duplicate-step.hook) to use new clipboard
  service
  - Updated usePasteEntity hook to use inferClipboardCompatibility() - determines compatibility by matching YAML
  key names against active canvas type
  - Updated DataMapper copy/paste/duplicate operations
  - Updated context menu item interaction helpers
  - Updated all visual entity classes to use IClipboardContent without the type field

  Test updates

  - Updated 167 lines of new clipboard service tests covering YAML parsing, serialization, and edge cases
  - Updated Cypress e2e tests to verify YAML structure instead of metadata
  - Updated all unit tests for visual entities, hooks, and DataMapper operations
  - Removed 193 lines of old ClipboardManager tests

  ---
  Why it changed

  User value

  - Enables copying entities as human-readable YAML that can be shared, edited externally, and pasted back
  - Supports pasting YAML from external sources (documentation, chat, text editors) directly into the canvas
  - Improves cross-application workflows and knowledge sharing

  Technical improvements

  - Simpler compatibility model: Infers entity type from YAML structure (route:, pipe:, to:) instead of requiring
  metadata
  - Better separation of concerns: Dedicated service class instead of static utility
  - Standards-aligned: Uses proper YAML serialization with appropriate formatting for entities vs steps
  - More robust parsing: Handles both array and object YAML formats

  ---
  How it was verified

  Unit tests

  - ✅ All existing tests updated and passing
  - ✅ New clipboard.service.test.ts covers:
    - Parsing both array and object YAML formats
    - Serialization for entities (route, pipe, kamelet) and steps
    - MIME type detection and fallback behavior
    - Edge cases (invalid YAML, empty content, multiple keys)

  E2e tests

  - ✅ Updated Cypress tests verify:
    - Copy operations write correct YAML structure
    - Paste operations accept YAML without metadata
    - Step, route, and pipe copy/paste flows work end-to-end

  Manual testing checklist

  - [ ] Copy a route from canvas → verify YAML in clipboard is properly formatted
  - [ ] Copy a step from canvas → verify YAML includes leading dash for array format
  - [ ] Paste external YAML (e.g., route: { id: 'test', ... }) → verify it creates entity on canvas
  - [ ] Paste Kaoto YAML (e.g., [{ to: { uri: 'log:test' } }]) → verify it creates step
  - [ ] Test on both Chrome (with web text/kaoto) and Firefox (with text/plain)
  - [ ] Verify incompatible paste shows error modal

  ---
  Reviewer notes

  Key files to review

  1. packages/ui/src/services/visualization/clipboard.service.ts - New clipboard service with YAML serialization
  logic
  2. packages/ui/src/hooks/usePasteEntity.ts - Updated compatibility inference logic
  3. packages/ui/src/models/visualization/clipboard.ts - Simplified interface (removed type field)
  4. Cypress tests (stepCopy.cy.ts, stepPaste.cy.ts) - E2e verification

  Design decisions

  - Entity vs step formatting: Top-level entities (route, pipe, etc.) serialize as objects; steps serialize as
  single-element arrays to match YAML DSL conventions (leading dash)
  - Compatibility inference: Matches clipboard YAML key name against supportedEntities or resource type
  - Entity vs step formatting: Top-level entities (route, pipe, etc.) serialize as objects; steps serialize as single-element arrays to match YAML DSL conventions (leading dash)
  - Compatibility inference: Matches clipboard YAML key name against supportedEntities or resource type (case-insensitive) instead of using metadata
  - MIME type strategy: Chrome gets both web text/kaoto and text/plain; other browsers get text/plain only

  Backwards compatibility

  - ✅ All existing copy/paste functionality preserved
  - ✅ Internal clipboard operations continue to work
  - ⚠️  Old clipboard format (with type field) is no longer written, but paste compatibility is maintained via key name inference
  
  fix: https://github.com/KaotoIO/kaoto/issues/2652

## Summary by CodeRabbit

* **New Features**
  * Clipboard copy/paste now uses a YAML-based format with Kaoto-aware MIME support for better interoperability with external editors.
* **Bug Fixes**
  * Pasting steps from raw YAML (without any app-specific marker) works reliably.
  * Simplified clipboard payloads and updated compatibility checks make paste behavior align with supported components.
* **Chores**
  * Updated automated UI and unit tests to match the new clipboard format and parsing behavior.

### CODE REVIEW

**Review by igarashitm — COMMENTED** [review]
No issue is found from DataMapper PoV 👍

File: packages/ui/src/main.tsx [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import './components/SourceCode/workers/enable-workers'; // Configure Monaco loader before anything else
```
Comment (lordrip) [inline packages/ui/src/main.tsx]: Interesting, why is this? 👀

File: packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts [hunk 2]
```diff
@@ -87,18 +87,18 @@ describe('Tests for Design page', { browser: '!firefox' }, () => {
...
-  it('Design - Paste steps in CamelRoute by adding something to the clipboad', () => {
+  it('Design - Paste steps in CamelRoute by adding something to the clipboard', () => {
     cy.uploadFixture('flows/camelRoute/basic.yaml');
-    cy.addValueToClipboard({
-      type: 'Route',
-      name: 'to',
-      definition: { id: 'to-1913', uri: 'amqp', parameters: {} },
-      __kaoto_marker: 'kaoto-node',
+    cy.window().then(async (win) => {
+      const yamlContent = '- to:\n    id: to-1913\n    uri: amqp\n    parameters: {}\n';
```
Comment (lordrip) [inline packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts]:
```suggestion
      const yamlContent = ```
- to:
    id: to-1913
    uri: amqp
    parameters: {}
```;

File: packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts [hunk 3]
```diff
@@ -108,4 +108,26 @@ describe('Tests for Design page', { browser: '!firefox' }, () => {
       expect(uriMatches).to.have.lengthOf(1);
     });
   });
+
+  it('Design - Paste step from external YAML source (no Kaoto marker)', () => {
+    cy.uploadFixture('flows/camelRoute/basic.yaml');
+    // Simulate clipboard content written by an external editor — raw YAML, no __kaoto_marker
+    cy.window().then(async (win) => {
+      const externalYaml = '- log:\n    message: hello from external\n';
```
Comment (lordrip) [inline packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts]:
```suggestion
      const externalYaml = ```
- log:
    message: hello from external
```;

**Review by lordrip — COMMENTED** [review]
I tried it locally and it works very well. Considering that we're reading/writing XLM from/to the clipboard, I think we should move away from using markers and custom serialization and instead use plain `string`. XML support should be added in an upcoming PR.

File: packages/ui/src/main.tsx (hunk 1 above)
Comment (PVinaches) [reply packages/ui/src/main.tsx]: It was a Claude suggestion and I digged a bit and what this does is to load Monaco first so that it avoids falling into the CDN based loading. If for ex. it uses the CDN one, we could have version incompatibility at some point or possible failures. I think it's worth trying out. In my end I didn't notice any issue so I decided to keep the change here in case we can enjoy some benefits also in the E2E experience.

File: packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts (hunk 3 above)
Comment (PVinaches) [reply packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts]: pushed with just one `

File: packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts (hunk 2 above)
Comment (PVinaches) [reply packages/ui-tests/cypress/e2e/designer/basicNodeActions/stepPaste.cy.ts]: pushed with just one `

File: packages/ui/src/main.tsx (hunk 1 above)
Comment (PVinaches) [reply packages/ui/src/main.tsx]: I deleted this file from the PR as it doesn't influence directly tnis. But I think we should think about this in relation of those random E2E fails

**Review by lordrip — APPROVED** [review]
